transformer_engine/pytorch/attention.py

```diff
@@ -529,36 +555,25 @@ def forward(
         """
 
         use_flash_attention = self.use_flash_attention
-        if (query_layer.dtype not in [torch.bfloat16, torch.float16]
+        if (query_layer.dtype not in [torch.bfloat16, torch.float16] # pylint: disable=too-many-boolean-expressions
             or key_layer.dtype not in [torch.bfloat16, torch.float16]
             or value_layer.dtype not in [torch.bfloat16, torch.float16]
             or (self.device_compute_capability == 8.6 and key_layer.shape[-1] > 64)
+            or is_in_onnx_export_mode()
         ):
             use_flash_attention = False
 
-        if self.attn_mask_type == "padding" and attention_mask is not None:
-            use_flash_attention = False
-
-        if is_in_onnx_export_mode():
-            use_flash_attention = False
-
-        if use_flash_attention:
-            if checkpoint_core_attention:
-                return self._checkpointed_attention_forward(self.flash_attention,
-                                                            query_layer,
-                                                            key_layer,
-                                                            value_layer)
-            return self.flash_attention(query_layer, key_layer, value_layer)
+        attention_func = self.flash_attention if use_flash_attention else self.unfused_attention
```

Comment:
Should we include self.fused_attention in this logic as well? In my PR, I kept the selection order as flash -> fused -> unfused.

Reply:
Yes I like this hierarchy. I'm switching to it as I resolve all the conflicts